packages/gatsby/src/schema/infer/inference-metadata.js

```diff
@@ -118,55 +153,124 @@ const getType = (value, key) => {
+const updateValueDescriptorObject = (
+  value,
+  typeInfo,
+  nodeId,
+  operation,
+  metadata,
+  path
 ) => {
-  // The object may be traversed multiple times from root.
-  // Each time it does it should not revisit the same node twice
-  if (path.includes(value)) {
-    return [descriptor, false]
-  }
+  path.push(value)
 
-  const typeName = getType(value, key)
+  const { dprops = {} } = typeInfo
+  typeInfo.dprops = dprops
 
-  if (typeName === `null`) {
-    return [descriptor, false]
-  }
+  Object.keys(value).forEach(key => {
+    const v = value[key]
 
-  path.push(value)
+    let descriptor = dprops[key]
+    if (descriptor === undefined) {
+      dprops[key] = descriptor = {}
+    }
 
-  const ret = _updateValueDescriptor(
-    nodeId,
-    key,
-    value,
-    operation,
-    descriptor,
-    path,
-    typeName
-  )
+    updateValueDescriptor(nodeId, key, v, operation, descriptor, metadata, path)
+  })
 
   path.pop()
+}
+
+const updateValueDescriptorArray = (
+  value,
+  key,
+  typeInfo,
+  nodeId,
+  operation,
+  metadata,
+  path
+) => {
+  value.forEach(item => {
+    let descriptor = typeInfo.item
+    if (descriptor === undefined) {
+      typeInfo.item = descriptor = {}
+    }
+
+    updateValueDescriptor(
+      nodeId,
+      key,
+      item,
+      operation,
+      descriptor,
+      metadata,
+      path
+    )
+  })
+}
+
+const updateValueDescriptorRelNodes = (
+  listOfNodeIds,
+  delta,
+  operation,
+  typeInfo,
+  metadata
+) => {
+  const { nodes = {} } = typeInfo
+  typeInfo.nodes = nodes
+
+  listOfNodeIds.forEach(nodeId => {
+    nodes[nodeId] = (nodes[nodeId] || 0) + delta
+
+    // Treat any new related node addition or removal as a structural change
+    // FIXME: this will produce false positives as this node can be
+    //  of the same type as another node already in the map (but we don't know it here)
+    if (nodes[nodeId] === 0 || (operation === `add` && nodes[nodeId] === 1)) {
+      metadata.dirty = true
+    }
+  })
+}
 
-  return ret
+const updateValueDescriptorString = (value, delta, typeInfo) => {
+  if (value === ``) {
+    const { empty = 0 } = typeInfo
+    typeInfo.empty = empty + delta
+  }
+  typeInfo.example =
+    typeof typeInfo.example !== `undefined` ? typeInfo.example : value
 }
-const _updateValueDescriptor = (
+
+const updateValueDescriptor = (
   nodeId,
   key,
   value,
-  operation,
+  operation = `add` /* add | del */,
   descriptor,
-  path,
-  typeName
+  metadata,
+  path
 ) => {
+  // The object may be traversed multiple times from root.
+  // Each time it does it should not revisit the same node twice
+  if (path.includes(value)) {
+    return
+  }
+
+  const typeName = getType(value, key)
+
+  if (typeName === `null`) {
+    return
+  }
+
   const delta = operation === `del` ? -1 : 1
-  const typeInfo = descriptor[typeName] || { total: 0 }
+  let typeInfo = descriptor[typeName]
+  if (typeInfo === undefined) {
+    typeInfo = descriptor[typeName] = { total: 0 }
+  }
```

Comment:
Can we avoid double assignments for readability? I used them before but found they are confusing for many people. Maybe just do:
```js
  if (descriptor[typeName] === undefined) {
    descriptor[typeName] = { total: 0 }
  }
  const typeInfo = descriptor[typeName]
```

Reply:
Yeah you're right. I shouldn't have gone for it.